src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -8,11 +8,1582 @@
 using NHibernate.Util;
 using Remotion.Linq;
 using Remotion.Linq.Parsing.ExpressionTreeVisitors;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace NHibernate.Linq
 {
 	public static class LinqExtensionMethods
 	{
+		private static readonly Dictionary<string, MethodInfo> cachableQueryableMethods;
+
+		static LinqExtensionMethods()
+		{
+			cachableQueryableMethods = new Dictionary<string, MethodInfo>
+			{
+				{"Count", GetMethod(() => new EnumerableQuery<int>(new int[] { }).Count())},
```

Comment:
Should use `ReflectHelper.GetMethodDefinition<IQueryable<int>>(q => q.Count())`.

`QueryableGenerator` already resolves some of these, maybe should they be put in `ReflectionCache`. 

And for avoiding using "magic strings", maybe a pattern similar to the `ReflectionCache` should be used here.